intermediate_source/pipelining_tutorial.rst

```diff
@@ -0,0 +1,213 @@
+         split_spec={
+            "layers.4": SplitPoint.BEGINNING,
+         }
+      )
+      stage = pipe.build_stage(stage_index, device, pp_group)
+      return stage
+
+
+Step 2: Define The Main Execution
+---------------------------------
+
+In the main function we will create a particular pipeline schedule that the stages should adhere to. `torch.distributed.pipelining`
+supports multiple schedules including single stage per rank schedules GPipe and 1F1B and multiple stage per rank schedules such as
+Interleaved1F1B and LoopedBFS.
+
+.. code:: python
+
+   if __name__ == "__main__":
+      init_distributed()
+      num_microbatches = 4
+      model_args = ModelArgs()
+      model = Transformer(model_args)
+
+      # Dummy data
+      x = torch.ones(32, 500, dtype=torch.long)
+      y = torch.randint(0, model_args.vocab_size, (32, 500), dtype=torch.long)
+      example_input_microbatch = x.chunk(num_microbatches)[0]
+
+      # Option 1: Manual model splitting
+      stage = manual_model_split(model, example_input_microbatch, model_args)
+
+      # Option 2: Tracer model splitting
+      # stage = tracer_model_split(model, example_input_microbatch)
+
+      x = x.to(device)
+      y = y.to(device)
+
+      def tokenwise_loss_fn(outputs, targets):
+         loss_fn = nn.CrossEntropyLoss()
+         outputs = outputs.view(-1, model_args.vocab_size)
+         targets = targets.view(-1)
+         return loss_fn(outputs, targets)
+
+      schedule = ScheduleGPipe(stage, n_microbatches=num_microbatches, loss_fn=token_loss_fn)
+
+      if rank == 0:
+         schedule.step(x)
+      elif rank == 1:
+         losses = []
+         output = schedule.step(target=y, losses=losses)
+      dist.destroy_process_group()
+
+We are using the manual option of splitting the model, but the code can be uncommented to also try the
+tracer based model splitting function. In our schedule we need to pass in the number of microbatches and
+the loss function we are using to evaluate the targets.
+
+The `.step()` function runs the entire minibatch and performs automatic splitting into microbatches based
+on the `n_microbatches` passed previously. The microbatches are then operated on according to the schedule class.
+In this example we are using GPipe, so it is a simple all forwards and then all backwards schedule. The output
+returned from rank 1 will be the same as if the model was on a single GPU and run with the entire batch. Similarly,
+we can pass in a `losses` container to store the corresponding losses for each microbatch.
+
+Step 3: Launch the Distributed Processes
+----------------------------------------
+
+Finally, we are ready to run the script. We will use `torchrun` to create a single host, 2 process job.
+Our script is already written in a way rank 0 will performs the required logic for pipeline stage 0 and rank 1
+will performs the logic for pipeline stage 1.
+
+`torchrun --standalone --nnodes 1 --nproc_per_node 2 pipelining_tutorial.py`
```

Comment:
Can we please add a conclusion? Something like:

Conclusion
-----------

In this tutorial, we have learned how to implement distributed pipeline parallelism using PyTorch's ``torch.distributed.pipelining`` APIs. We explored setting up the environment, defining a transformer model, and partitioning it for distributed training. We discussed two methods of model partitioning—manual and tracer-based—and demonstrated how to schedule computations on micro-batches across different stages. Finally, we covered the execution of the pipeline schedule and the launch of distributed processes using ``torchrun``. This guide showcased the effectiveness of pipeline parallelism in training large models efficiently.

See also:

* Link1
* Link2
* ...

Reply:
Awesome, thanks for writing that up. Added it